### PR TITLE
Amazon crawling

### PR DESCRIPTION
**Pull request by maiphuong-van**

### CODE REVIEW

File: config/config.exs [hunk 1]
```diff
@@ -1,9 +1,25 @@
 use Mix.Config
 
+config :erlang_node_discovery,
+  hosts: ["127.0.0.1", "crawlyui.com"],
+  node_ports: [{:ui, 0}]
+
+ui_node = System.get_env("UI_NODE") || "ui@127.0.0.1"
+ui_node = ui_node |> String.to_atom()
+
```
Comment (oltarasenko) [inline config/config.exs]: This does not seem to be related to  the spider work itself.

File: config/config.exs [hunk 2]
```diff
@@ -1,9 +1,25 @@
...
+    Crawly.Middlewares.RobotsTxt,
+    {Crawly.Middlewares.UserAgent, user_agents: ["Crawly bot 1", "Crawly bot 2"]}
+  ],
+  # fetcher: {Crawly.Fetchers.Splash, [base_url: "http://localhost:8050/render.html"]},
   pipelines: [
-    {Crawly.Pipelines.Validate, fields: [:title, :location]},
-    CrawlyElixirJobs.Pipleline.Filter,
+    {Crawly.Pipelines.Validate, fields: [:id]},
+    {Crawly.Pipelines.DuplicatesFilter, item_id: :id},
+    {Crawly.Pipelines.Experimental.SendToUI, ui_node: ui_node},
```
Comment (oltarasenko) [inline config/config.exs]: Not needed as I see.

File: lib/crawly_elixir_jobs/middleware/amazon_unique_request.ex [hunk 3]
```diff
@@ -0,0 +1,36 @@
+defmodule CrawlyExlirJobs.MiddleWare.AmazonUniqueRequest do
+  # try to filter out request to same page on Amazon but does not entirely match existing request:
+  # ie "https://www.amazon.de/s?k=Drawing&page=3&qid=1598532609&ref=sr_pg_3" and "https://www.amazon.de/s?k=Drawing&page=3&qid=1598532611&ref=sr_pg_3"
+
+  require Logger
+
+  def run(request, state) do
+    unique_request_seen_requests = Map.get(state, :unique_request_seen_requests, [])
+
+    case Enum.any?(unique_request_seen_requests, &match_request(&1, request)) do
```
Comment (oltarasenko) [inline lib/crawly_elixir_jobs/middleware/amazon_unique_request.ex]: This is going to slow us down quite a bit. The number of seen requests is going to be huge. I would prefer to visit several duplicates and to get more products. 

Another option is to avoid pages by regexp in URLs..

File: config/config.exs (hunk 1 above)
Comment (maiphuong-van) [reply config/config.exs]: No thats' true, that was left over when I tried playing with the Crawly UI

File: config/config.exs (hunk 2 above)
Comment (maiphuong-van) [reply config/config.exs]: 👍

File: lib/crawly_elixir_jobs/middleware/amazon_unique_request.ex (hunk 3 above)
Comment (maiphuong-van) [reply lib/crawly_elixir_jobs/middleware/amazon_unique_request.ex]: I will try the regex suggestion :) I want to filter out as much duplicate as possible so I could also do more other requests to amazon before they don't like me anymore :D

File: config/config.exs (hunk 1 above)
Comment (oltarasenko) [reply config/config.exs]: I would prefer to follow more link, comparing to the idea of querying all seen URLs, as it's expensive. I would imagine a full amazon crawl would require you to make 100M of requests at least.

Comment (maiphuong-van) [reply config/config.exs]: Hmm I see, yeah in that sense it would be better to follow more links. I guess I need to optimise more on how to define user agents and request to amazon so it won't block me before I get to the 100M requests haha
